File: packages/typescript/aws/src/stack.ts
import { BaseStack, StackGroup } from '@awslv/cdk-organizer-core';
import { Stack as AwsStack, StackProps, Tags } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as _ from 'lodash';

export class Stack<TStackProps extends StackProps> extends AwsStack {
  private base: BaseStack;
  public config: Record<string, unknown>;
  public env: string;
  getResourceName: typeof StackGroup.prototype.getResourceName;
  getBucketName: typeof BaseStack.prototype.getBucketName;

  constructor(
    stackGroup: StackGroup,
    scope?: Construct,
    id?: string,
    props?: TStackProps
  ) {
    super(scope, id, {
      ...props,
      env: {
        account: stackGroup.data.account,
        region: stackGroup.data.region,
      },
    });

    this.base = new BaseStack(stackGroup);
    this.config = stackGroup.config;
    this.env = stackGroup.env;

    this.getResourceName = this.base.getResourceName.bind(this.base);
    this.getBucketName = this.base.getBucketName.bind(this.base);

    Tags.of(this).add('Environment', this.base.config['env'] as string);
    if (!_.isNil(this.base.config['tags'])) {
      Object.entries(
        this.base.config['tags'] as Record<string, string>
      ).forEach(([key, value]) => {
        Tags.of(this).add(key, value);
      });
    }
  }
}

File: packages/python/cdk_organizer/aws/stack.py
"""CDK Infra Core AWS Base Stack."""

from typing import TYPE_CHECKING

import aws_cdk as cdk

if TYPE_CHECKING:
    from constructs import Construct
    from cdk_organizer.aws.stack_group import StackGroup

from cdk_organizer.decorators.catch_exceptions import catch_exceptions
from cdk_organizer.stack import BaseStack


class Stack(BaseStack, cdk.Stack):
    """
    Base class for AWS CDK Stacks.

    Args:
        scope (Construct): AWS CDK Construct object
        id (str): Stack Id
        stack_group (StackGroup): StackGroup instance
    """

    @catch_exceptions
    def __init__(
        self,
        scope: "Construct",
        id: str,
        stack_group: "StackGroup",
        **kwargs
    ) -> None:
        """Initialize the class."""
        BaseStack.__init__(self, id, stack_group)
        cdk.Stack.__init__(self, scope, id, env=self.env_props, **kwargs)

        cdk.Tags.of(self).add('Environment', self.env_name.upper())
        if 'tags' in self.config:
            for key, value in self.config['tags'].items():
                cdk.Tags.of(self).add(key, value)

        self.node.add_metadata('cdk:module', self.stack_group.module_name)
